### ISSUE TITLE
Issue with Canadian French ordinal 

### ISSUE THREAD

**Issue by smilehyun** [moment/moment#3375]
Hi, 
i know only small portion of people would use French Canadian, such as Quebec, but found a bug in ordinal.

in fr.js

```
ordinal : function (number) {
     return number + (number === 1 ? 'er' : '');
}
```

but in fr-ca.js,

```
ordinal : function (number) {
      return number + (number === 1 ? 'er' : 'e');
}
```

Ordinal should be same as French, which is 'er' only after number 1.

Thanks

**Comment by sly7-7** [comment]
taken from http://www.academie-francaise.fr/abreviations-des-adjectifs-numeraux it seems like the issue ia the `e` is missing in the fr locale.

**Comment by fbonzon** [comment]
I noticed this issue too, and would like to prepare a PR to fix it.

In this thread and all previous about the same subject (like #2414 mentioned by @sly7-7), there is a confusion about **what** are the rules to write an ordinal in French in general without any context, and **when** to write an ordinal in the specific context of the day of the month.

In moment library, the `ordinal` element of a locale is not specific to the day of the month. It is used in many tokens:
Month `Mo` 1st 2nd ... 11th 12th
Quarter `Qo` 1st 2nd 3rd 4th
Day of Month `Do` 1st 2nd ... 30th 31st
Day of Year `DDDo` 1st 2nd ... 364th 365th
Day of Week `do` 0th 1st ... 5th 6th
Week of Year `wo` 1st 2nd ... 52nd 53rd
Week of Year (ISO) `Wo` 1st 2nd ... 52nd 53rd

When speaking or writing dates in English, both cardinal figures (July 4, format `D`) and ordinal figures (July 4th, format `Do`) are possible for every day of the month, depending on context, intended style, or English country sublocale. A given date format will use `D` or `Do` for every day of the month.

When speaking or writing dates in French, the rule is always the same: use ordinal for day 1 (suffix "er", format `Do`), and cardinal (format `D`) for other days. Day 1 is really always pronounced as an ordinal (1er = premier), but other days as cardinals (2 = deux, 3 = trois, etc.)

There is currently a hack in the 'fr' locale, where `ordinal` is defined with 'er' suffix for number 1, and with no suffix for other numbers. This makes using `Do` in date formats produce the expected results, but breaks all the others formats using an ordinal (see above: `Mo`, `Qo`, `DDDo`, `do`, `wo`, `Wo`).

This should be fixed. Use "e" suffix for numbers 2 and above, like mentioned in the link to Académie Française (authority for French rules) by @sly7-7 in the previous message.

Now, we need a way to express that long date formats in French need ordinal for day 1 and cardinal for other days. There is currently no mechanism to achieve this without adding some JavaScript code.

I see 2 solutions:

1. Like is already done for `relativeTime` element (and few others), e.g. in Finnish, allow to pass a function to the elements of `longDateFormat`, with an additional parameter for day number. We would then be able to return, for example:
`Do MMMM YYYY` when day = 1
`D MMMM YYYY` otherwise

2. [My preferred solution] Define a new token, say `DL` (for day localized), that is like `Do` by default, but with localized rules to include or not the ordinal suffix. Use this new token in the French locale date formats, for example: `LL : 'DL MMMM YYYY'`

Please advise what is your preferred solution. Can you think about another one too? (Adding to the discussion authors of the fr locales: @jfroffice @gaspard @jonbca )

**Comment by gaspard** [comment]
Thanks for notifying. I think solution (2) is fine as long as, like you said, it resolves to Do when nothing special is done.

I think calling it "Dom", making it clear what it is "Day of month" or "Day ordinal for month" is better then DL because everything is "localized" so the "L" does not add meaning...

**Comment by jfroffice** [comment]
@fbonzon solution (2) seems quite better as it is not disruptive with the current implementation.

+1 for "Dom" as @gaspard suggest.

**Comment by fbonzon** [comment]
I tried to implement 1., but abandoned. Is possible but requires more work than expected. The parsing code expects fixed formats, and allowing functions for the date formats (instead of fixed strings) incurs more changes than I can afford now. I liked that solution, because I think semantically this issue is better handled in functions within the French locale code, rather than at a global scope.

About 2., in my opinion `Dom` should not allow the possibility to stand for "Day ordinal for month". It unfortunately brings confusion, because it will not necessarily be an ordinal. It will be either an ordinal or a cardinal. We could also simply pick an unused letter. Say `F`, because it's the one after `D`, `E` being already taken.

**Comment by fbonzon** [comment]
Can a team member from moment confirm option 2 is OK (add a new token for Localized Day of month), before I start coding it ? Or feel free to claim this work. Thank you.

**Comment by marwahaha** [comment]
Moment.js will not take more substantial changes to its codebase.
https://momentjs.com/docs/#/-project-status/

it seems like @fbonzon 's changes require a lot of work, so I don't think we should proceed.

**Comment by mrleblanc101** [comment]
We say `Samedi 1er avril`, but not `Dimanche 2e avril`. 
The correct for should be `Dimanche 2 avril`.
I guess I'll override the translation myself in my projet to bring back the previous behavior which was the correct one.
